src/twisted/internet/abstract.py

```diff
@@ -269,27 +273,31 @@ def doWrite(self):
         if self.offset == len(self.dataBuffer) and not self._tempDataLen:
             self.dataBuffer = b""
             self.offset = 0
-            # stop writing.
-            self.stopWriting()
             # If I've got a producer who is supposed to supply me with data,
             if self.producer is not None and (
                 (not self.streamingProducer) or self.producerPaused
             ):
                 # tell them to supply some more.
                 self.producerPaused = False
+                # The presumption is that this will write some data, so as an
+                # optimization we don't need to call stopWriting():
                 self.producer.resumeProducing()
```

Comment:
I find the code from this method a bit hard to read / follow.

Can we return here ?

This should reduce the need for the final `else` ... and I hope will make the easier to read.

That is, conver each `elif CONDITIONT` to just `if CONDITION + return at the end` 

```suggestion
                # optimization we don't need to call stopWriting():
                self.producer.resumeProducing()
                return
```